### DISCUSSION TITLE
StubLSP: LSP for languages without LSP that treats file as plain text

### DISCUSSION THREAD

**Discussion by IvanGoncharov** [helix-editor/helix#6731] [Ideas]
Helix splits my UX into two categories:
- Popular file formats 👍 Everything is working great with minimal setup.
- Less popular file formats (e.g. CSV, git commit messages, TXT files) 👎 No advanced features.

I need autocomplete for my workflow that works without LSP: https://github.com/helix-editor/helix/pull/3328#issuecomment-1500209812
So I'm stuck doing local builds of #3328 that @estin heroically rebases for almost a year.
Moreover, I see other people trying to push different features to work without LSP.
Today I saw yet another one here: https://github.com/helix-editor/helix/issues/3436#issuecomment-1215009898

I sympathize with a Helix developer's attempt to keep the architecture simple.
And it's great that all IDE-like features are delegated to LSP and implementing fallbacks for every feature increase core size.

So I have this idea: Can we have a stub LSP that is by default applied to all languages without LSP?
That stub LSP can implement all LSP-like features but treat input as plain text.
In situations where TreeSitter grammar is available, it can make StubLSP a little bit smarter (e.g. distinguish between keywords and identifiers).

In an ideal world, I would like specialized LSP plugins to existing for every file format I used
but realistically speaking StubLSP would be a great DX boost for me today.

So my question for Helix devs:
1. Do you think something like this can be a good long-term solution to support file formats without LSP?
2. Can something like StubLSP be implemented as an external repo without losing any functionality?

P.S. I would love to work on StubLSP myself, but I'm overburdened with my own OSS projects and also have zero Rust knowledge.
So this idea along with the `StubLSP` name can be used by anyone who wants to build PoC.

**Comment by nluetts** [comment]
Have a look at https://github.com/helix-editor/helix/pull/3328#issuecomment-1591004675
